include/xtensor/xshape.hpp

```diff
@@ -333,6 +334,37 @@ namespace xt
 
     template <class S>
     using index_from_shape_t = typename detail::index_from_shape_impl<S>::type;
+
+    /**
+     * Deduce the common value type of underlying containers
+     */
+    template <class... C>
+    struct common_value_type {
+        using type = std::common_type_t<typename std::decay_t<C>::value_type...>;
+    };
+    template <class... C>
+    using common_value_type_t = typename common_value_type<C...>::type;
```

Comment:
There is already a definition of `common_value_type_t` in `xfunction.hpp`, it could be nice to remove it and use this one so we don't have code duplication.

Reply:
I'm only seeing `common_size_type_t` and `common_difference_type_t` in `xfunction.hpp`, no `common_value_type_t`.

Reply:
@JohanMabille Would you rather I move this implementation into `xfunction.hpp`?

Reply:
`xfunction.hpp` already includes `xshape.hpp`, so the definition of `common_value_type` should be in `xshape.hpp` for now. Later we will have a `expression_traits` file where we can gather the metafunctions about expressions types only.

Reply:
@JohanMabille I don't see any implementation of `common_value_type` in `xfunction.hpp` on the master branch.

Reply:
Indeed it looks like it has been removed! Sorry for that, things are moving too fast ;)